java/src/main/java/com/logicalclocks/hsfs/engine/FeatureViewEngine.java

```diff
@@ -28,135 +45,53 @@
-    featureView.setFeatures(updatedFeatureView.getFeatures());
-    return featureView;
-  }
-
-  static List<TrainingDatasetFeature> makeLabelFeatures(Query query, List<String> labels) throws FeatureStoreException {
-    if (labels == null || labels.isEmpty()) {
-      return Lists.newArrayList();
-    } else {
-      // If provided label matches column with prefix, then attach label.
-      // If provided label matches only one column without prefix, then attach label. (For
-      // backward compatibility purpose, as of v3.0, labels are matched to columns without prefix.)
-      // If provided label matches multiple columns without prefix, then raise exception because it is ambiguous.
-
-      Map<String, String> labelWithPrefixToFeature = Maps.newHashMap();
-      Map<String, FeatureGroup> labelWithPrefixToFeatureGroup = Maps.newHashMap();
-      Map<String, List<FeatureGroup>> labelToFeatureGroups = Maps.newHashMap();
-      for (Feature feat : query.getLeftFeatures()) {
-        labelWithPrefixToFeature.put(feat.getName(), feat.getName());
-        labelWithPrefixToFeatureGroup.put(feat.getName(),
-            (new FeatureGroup(null, feat.getFeatureGroupId())));
-      }
-      for (Join join : query.getJoins()) {
-        for (Feature feat : join.getQuery().getLeftFeatures()) {
-          String labelWithPrefix = join.getPrefix() + feat.getName();
-          labelWithPrefixToFeature.put(labelWithPrefix, feat.getName());
-          labelWithPrefixToFeatureGroup.put(labelWithPrefix,
-              new FeatureGroup(null, feat.getFeatureGroupId()));
-          List<FeatureGroup> featureGroups = labelToFeatureGroups.getOrDefault(feat.getName(), Lists.newArrayList());
-          featureGroups.add(new FeatureGroup(null, feat.getFeatureGroupId()));
-          labelToFeatureGroups.put(feat.getName(), featureGroups);
-        }
-      }
-      List<TrainingDatasetFeature> trainingDatasetFeatures = Lists.newArrayList();
-      for (String label : labels) {
-        if (labelWithPrefixToFeature.containsKey(label)) {
-          trainingDatasetFeatures.add(new TrainingDatasetFeature(
-              labelWithPrefixToFeatureGroup.get(label.toLowerCase()),
-              labelWithPrefixToFeature.get(label.toLowerCase()),
-              true));
-        } else if (labelToFeatureGroups.containsKey(label)) {
-          if (labelToFeatureGroups.get(label.toLowerCase()).size() > 1) {
-            throw new FeatureStoreException(String.format(AMBIGUOUS_LABEL_ERROR, label));
-          }
-          trainingDatasetFeatures.add(new TrainingDatasetFeature(
-              labelToFeatureGroups.get(label.toLowerCase()).get(0),
-              label.toLowerCase(),
-              true));
-        } else {
-          throw new FeatureStoreException(String.format(LABEL_NOT_EXIST_ERROR, label));
-        }
-
-      }
-      return trainingDatasetFeatures;
-    }
-  }
-
 
   public FeatureView update(FeatureView featureView) throws FeatureStoreException,
```

Comment:
Move methods like this to the base file. They are framework agnostic. (there are more methods like this in this file)

Reply:
you have to move, `getOrCreateFeatureView`, `getBatchQuery`, `getBatchQueryString`, `deleteTrainingDatasetOnly`, `deleteTrainingDatasetOnly`, `deleteTrainingData`, `deleteTrainingData`, `recreateTrainingDataset`, `getTrainingDataMetadata`, `setEventTime`, `createTrainingDataMetadata`, `setTrainSplit`, `getTrainingDataset`,  `getTrainingDataset`, `createTrainingDataset`, `get`, `get`, `update`


add abstract method in base and @Overwrite for `getBatchData`, `readDataset`, `convertSplitDatasetsToMap`, `computeStatistics`, `castColumnType`, `getTrainingDataset`, `writeTrainingDataset`

Reply:
I move some stuff but not everything can be moved as you suggested. For `QueryBase.java` -> `Query.java` suggestion that we discussed worked fine. But here I can't apply the same method. For example I can't make `getBatchData ` abstract because I need to provide `FeatureView`. Again parameter types are problem and since `FeatureView` is not related to `FeatureViewEngine` I can't apply the same technique what we did in `QueryBase.java`

Reply:
Also I can't move `getOrCreateFeatureView` to base as it requires `FeatureView.FeatureViewBuilder` and `Query.java`

Reply:
done, made getOrCreateFeatureView abstract